How many late days are you using for this assignment? None

In [357]:
# Imports and pip installations (if needed)
import pandas as pd
from scipy import stats
import numpy as np

# Part 1: Load the dataset

In [358]:
# Load the given datasets
categorical_data = pd.read_csv("chronic_kidney_disease_categorical.csv")
numerical_data = pd.read_csv("chronic_kidney_disease_numerical.csv")

# Print the data
print(categorical_data)
print(numerical_data)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
425     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   
426     872034  2.0  1.0  abnormal  abnormal  notpresent     present  yes   
427     643007  NaN  NaN       NaN    normal  notpresent  notpresent  yes   
428     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
429     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm  cad appet   pe  ane  Target  
0    yes   no  good  yes   no     c

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:
In chronic_kidney_disease_numerical.csv, there are only numerical values for all the features. There is age, blood pressure, blood glucose random, blood urea, serum creatinine, sodium, potassium, hemoglobin, packed cell volume, wbcc, and rbcc. All of them are numerical examples but three examples of numerical columns would be age (Integer), blood pressure (Integer), and blood urea (Integer). There are no examples of categorical columns in this dataset file.

In chronic_kidney_disease_categorical.csv, there are only categorical values for all the features. There is albumin, sugar, red blood cells, pus cell, pus cell clumps, bacteria, hypertension, diabetes mellitus, coronary artery disease, appetite, pedal edema, anemia, Target.  All of them are categorical examples but three examples of categorical columns would be albumin (Categorical), bacteria (Categorical), and red blood cells (Categorical). There are no examples of numerical columns in this dataset file.

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [359]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
num_dupes = numerical_data.shape[0] - numerical_data['unique_id'].nunique()
if(num_dupes > 0):
    print(f"{num_dupes}")


# Drop these duplicate rows
numerical_data = numerical_data.drop_duplicates(subset=['unique_id'])

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
cat_dupes = categorical_data.shape[0] - categorical_data['unique_id'].nunique()
if(cat_dupes > 0):
    print(f"{cat_dupes}")


categorical_data = categorical_data.drop_duplicates(subset=['unique_id'])



30
30


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer:

In [360]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
combined_data = pd.merge(numerical_data, categorical_data, on='unique_id')


#Print the combined dataset
combined_data


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
3,764115,70.0,90.0,144.0,125.0,4.0,136.0,4.6,12.0,37.0,...,normal,notpresent,notpresent,yes,yes,no,poor,yes,no,ckd
4,240975,47.0,80.0,95.0,35.0,0.9,140.0,4.1,NaN,NaN,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
385,522912,56.0,80.0,165.0,55.0,1.8,NaN,NaN,13.5,40.0,...,normal,notpresent,notpresent,yes,yes,no,poor,yes,no,ckd
386,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,...,abnormal,notpresent,present,yes,yes,no,good,yes,no,ckd
387,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
388,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,...,abnormal,present,present,no,no,no,good,no,yes,ckd


## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [361]:
# Calculate the percentage of rows that contain atleast one missing value

missing = combined_data.isnull().any(axis=1).mean() * 100

# Print %
print(f"{missing:}%")

# Drop these rows from the dataset
combined = combined_data.dropna()


# Print the Dataset
print(combined)


60.76923076923077%
     unique_id   age     bp    bgr    bu   sc    sod  pot  hemo   pcv  ...  \
0       203694  38.0   80.0   99.0  19.0  0.5  147.0  3.5  13.6  44.0  ...   
1       938027  43.0   60.0  108.0  25.0  1.0  144.0  5.0  17.8  43.0  ...   
2       421471  37.0   60.0  111.0  35.0  0.8  135.0  4.1  16.2  50.0  ...   
7       109053  64.0   60.0  106.0  27.0  0.7  150.0  3.3  14.4  42.0  ...   
13      349892  59.0   70.0  424.0  55.0  1.7  138.0  4.5  12.6  37.0  ...   
..         ...   ...    ...    ...   ...  ...    ...  ...   ...   ...  ...   
379     486397  20.0   70.0  123.0  44.0  1.0  135.0  3.8  14.6  44.0  ...   
381     462892  36.0   80.0   85.0  16.0  1.1  142.0  4.1  15.6  44.0  ...   
386     475200  64.0  100.0  163.0  54.0  7.2  140.0  4.6   7.9  26.0  ...   
387     168242  48.0   60.0  112.0  44.0  1.2  142.0  4.9  14.5  44.0  ...   
388     474407  21.0   90.0  107.0  40.0  1.7  125.0  3.5   8.3  23.0  ...   

           pc         pcc          ba  htn  

## Part 3.4: Sort the dataset according to the Labels

In [362]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
combined = combined.sort_values(by="Target").reset_index(drop=True)

# Print the dataset
print(combined)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
1       137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
2       484175  53.0   90.0   70.0  107.0  7.2  114.0  3.7   9.5  29.0  ...   
3       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
4       532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
148     647552  61.0   70.0  120.0   29.0  0.7  137.0  3.5  17.4  52.0  ...   
149     125587  37.0   60.0  109.0   47.0  1.1  141.0  4.9  15.0  48.0  ...   
150     614376  57.0   80.0  119.0   17.0  1.2  135.0  4.7  15.4  42.0  ...   
151     853426  41.0   70.0  125.0   38.0  0.6  140.0  5.0  16.8  41.0  ...   
152     572140  55.0   80.0  104.0   28.0  0.9  142.0  4.8  17.3  52.0  ...   

           pc         pcc          ba  htn   dm  ca

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [363]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
combined = pd.get_dummies(combined, columns=categorical_columns)

# Print the dataset
combined

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,...,True,False,True,False,True,False,False,True,True,False
1,137148,73.0,100.0,295.0,90.0,5.6,140.0,2.9,9.2,30.0,...,False,True,False,True,True,False,True,False,True,False
2,484175,53.0,90.0,70.0,107.0,7.2,114.0,3.7,9.5,29.0,...,True,False,False,True,True,False,False,True,True,False
3,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,...,False,True,False,True,False,True,False,True,True,False
4,532520,60.0,90.0,105.0,53.0,2.3,136.0,5.2,11.1,33.0,...,True,False,True,False,True,False,True,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,647552,61.0,70.0,120.0,29.0,0.7,137.0,3.5,17.4,52.0,...,True,False,True,False,True,False,True,False,False,True
149,125587,37.0,60.0,109.0,47.0,1.1,141.0,4.9,15.0,48.0,...,True,False,True,False,True,False,True,False,False,True
150,614376,57.0,80.0,119.0,17.0,1.2,135.0,4.7,15.4,42.0,...,True,False,True,False,True,False,True,False,False,True
151,853426,41.0,70.0,125.0,38.0,0.6,140.0,5.0,16.8,41.0,...,True,False,True,False,True,False,True,False,False,True


In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: 

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [364]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']


combined = combined[np.abs(stats.zscore(combined[numerical_columns]) < 3).all(axis=1)]

# Print the dataset
combined

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,...,True,False,True,False,True,False,False,True,True,False
1,137148,73.0,100.0,295.0,90.0,5.6,140.0,2.9,9.2,30.0,...,False,True,False,True,True,False,True,False,True,False
2,484175,53.0,90.0,70.0,107.0,7.2,114.0,3.7,9.5,29.0,...,True,False,False,True,True,False,False,True,True,False
3,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,...,False,True,False,True,False,True,False,True,True,False
4,532520,60.0,90.0,105.0,53.0,2.3,136.0,5.2,11.1,33.0,...,True,False,True,False,True,False,True,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,647552,61.0,70.0,120.0,29.0,0.7,137.0,3.5,17.4,52.0,...,True,False,True,False,True,False,True,False,False,True
149,125587,37.0,60.0,109.0,47.0,1.1,141.0,4.9,15.0,48.0,...,True,False,True,False,True,False,True,False,False,True
150,614376,57.0,80.0,119.0,17.0,1.2,135.0,4.7,15.4,42.0,...,True,False,True,False,True,False,True,False,False,True
151,853426,41.0,70.0,125.0,38.0,0.6,140.0,5.0,16.8,41.0,...,True,False,True,False,True,False,True,False,False,True


## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [365]:
# Normalize the all Numerical Attributes in the dataset.

from sklearn.preprocessing import MinMaxScaler


scaler = MinMaxScaler()
combined[numerical_columns] = scaler.fit_transform(combined[numerical_columns])

# Print the dataset
print(combined)


     unique_id       age    bp       bgr        bu        sc       sod  \
0       474407  0.194805  0.75  0.158798  0.196078  0.160494  0.358974   
1       137148  0.870130  1.00  0.965665  0.522876  0.641975  0.743590   
2       484175  0.610390  0.75  0.000000  0.633987  0.839506  0.076923   
3       343710  0.714286  0.50  0.442060  0.901961  0.432099  0.615385   
4       532520  0.701299  0.75  0.150215  0.281046  0.234568  0.641026   
..         ...       ...   ...       ...       ...       ...       ...   
148     647552  0.714286  0.25  0.214592  0.124183  0.037037  0.666667   
149     125587  0.402597  0.00  0.167382  0.241830  0.086420  0.769231   
150     614376  0.662338  0.50  0.210300  0.045752  0.098765  0.615385   
151     853426  0.454545  0.25  0.236052  0.183007  0.024691  0.743590   
152     572140  0.636364  0.50  0.145923  0.117647  0.061728  0.794872   

          pot      hemo       pcv  ...  cad_no  cad_yes  appet_good  \
0    0.256410  0.221311  0.184211  ...  

/var/folders/w0/0vsxrnfs6cg802b64pp6tf2h0000gn/T/ipykernel_64659/1211922613.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  combined[numerical_columns] = scaler.fit_transform(combined[numerical_columns])


## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer:

In [366]:
#Remove that column
# take out all unneccessary binary columns!!
remove = ['pe_no', 'ane_no', 'Target_notckd', 'cad_no',  'appet_poor', 'dm_no', 'htn_no', 'ba_notpresent', 'pcc_notpresent', 'pc_abnormal', 'rbc_abnormal']
combined.drop(columns=remove,inplace=True)

# Print the dataset
combined


/var/folders/w0/0vsxrnfs6cg802b64pp6tf2h0000gn/T/ipykernel_64659/584476101.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  combined.drop(columns=remove,inplace=True)


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,pc_normal,pcc_present,ba_present,htn_yes,dm_yes,cad_yes,appet_good,pe_yes,ane_yes,Target_ckd
0,474407,0.194805,0.75,0.158798,0.196078,0.160494,0.358974,0.256410,0.221311,0.184211,...,False,True,True,False,False,False,True,False,True,True
1,137148,0.870130,1.00,0.965665,0.522876,0.641975,0.743590,0.102564,0.295082,0.368421,...,False,True,False,True,True,True,False,False,False,True
2,484175,0.610390,0.75,0.000000,0.633987,0.839506,0.076923,0.307692,0.319672,0.342105,...,False,True,False,True,True,False,False,False,True,True
3,343710,0.714286,0.50,0.442060,0.901961,0.432099,0.615385,0.692308,0.172131,0.210526,...,False,False,False,True,True,True,False,True,True,True
4,532520,0.701299,0.75,0.150215,0.281046,0.234568,0.641026,0.692308,0.450820,0.447368,...,True,False,False,False,False,False,True,False,False,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,647552,0.714286,0.25,0.214592,0.124183,0.037037,0.666667,0.256410,0.967213,0.947368,...,True,False,False,False,False,False,True,False,False,False
149,125587,0.402597,0.00,0.167382,0.241830,0.086420,0.769231,0.615385,0.770492,0.842105,...,True,False,False,False,False,False,True,False,False,False
150,614376,0.662338,0.50,0.210300,0.045752,0.098765,0.615385,0.564103,0.803279,0.684211,...,True,False,False,False,False,False,True,False,False,False
151,853426,0.454545,0.25,0.236052,0.183007,0.024691,0.743590,0.641026,0.918033,0.657895,...,True,False,False,False,False,False,True,False,False,False


## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [367]:
# Export the dataframe to a new csv file
combined.to_csv("cleaned_chronic_kidney_disease.csv", index=False)

# Export the dataframe to a new json file
combined.to_json("cleaned_chronic_kidney_disease.json", orient="records")

print(combined.head(15))



    unique_id       age    bp       bgr        bu        sc       sod  \
0      474407  0.194805  0.75  0.158798  0.196078  0.160494  0.358974   
1      137148  0.870130  1.00  0.965665  0.522876  0.641975  0.743590   
2      484175  0.610390  0.75  0.000000  0.633987  0.839506  0.076923   
3      343710  0.714286  0.50  0.442060  0.901961  0.432099  0.615385   
4      532520  0.701299  0.75  0.150215  0.281046  0.234568  0.641026   
6      546225  0.519481  0.00  0.399142  0.535948  0.358025  0.769231   
8      514721  0.649351  0.75  0.253219  0.633987  0.777778  0.512821   
9      899916  1.000000  0.25  0.137339  0.326797  0.271605  0.102564   
10     621332  0.688312  0.50  1.000000  0.163399  0.111111  0.282051   
13     481293  0.701299  0.00  0.935622  0.169935  0.160494  0.487179   
15     297893  0.805195  0.50  0.373391  0.522876  0.456790  0.487179   
17     615697  0.000000  0.00  0.103004  0.372549  0.074074  0.615385   
18     337834  0.545455  0.25  0.201717  0.300654  

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

Paste here: 

| unique_id |       age |   bp |      bgr |      bu |       sc |       sod |       pot |      hemo |       pcv |   ... | pc_normal | pcc_present | ba_present | htn_yes | dm_yes | cad_yes | appet_good | pe_yes | ane_yes | Target_ckd |
|-----------|-----------|------|----------|---------|----------|-----------|-----------|-----------|-----------|-------|-----------|-------------|------------|---------|--------|---------|------------|--------|---------|------------|
|     474407 | 0.194805 | 0.75 | 0.158798 | 0.196078 | 0.160494 | 0.358974  | 0.256410  | 0.221311  | 0.184211  | ...   | False     | True        | True       | False   | False  | False   | True       | False  | True    | True       |
|     137148 | 0.870130 | 1.00 | 0.965665 | 0.522876 | 0.641975 | 0.743590  | 0.102564  | 0.295082  | 0.368421  | ...   | False     | True        | False      | True    | True   | True    | False      | False  | False   | True       |
|     484175 | 0.610390 | 0.75 | 0.000000 | 0.633987 | 0.839506 | 0.076923  | 0.307692  | 0.319672  | 0.342105  | ...   | False     | True        | False      | True    | True   | False   | False      | False  | True    | True       |
|     343710 | 0.714286 | 0.50 | 0.442060 | 0.901961 | 0.432099 | 0.615385  | 0.692308  | 0.172131  | 0.210526  | ...   | False     | False       | False      | True    | True   | True    | False      | True   | True    | True       |
|     532520 | 0.701299 | 0.75 | 0.150215 | 0.281046 | 0.234568 | 0.641026  | 0.692308  | 0.450820  | 0.447368  | ...   | True      | False       | False      | False   | False  | False   | True       | False  | True    | True       |
|     546225 | 0.519481 | 0.00 | 0.399142 | 0.535948 | 0.358025 | 0.769231  | 0.384615  | 0.344262  | 0.315789  | ...   | True      | False       | False      | True    | True   | False   | True       | False  | True    | True       |
|     514721 | 0.649351 | 0.75 | 0.253219 | 0.633987 | 0.777778 | 0.512821  | 0.589744  | 0.286885  | 0.342105  | ...   | False     | False       | False      | True    | False  | False   | True       | False  | True    | True       |
|     899916 | 1.000000 | 0.25 | 0.137339 | 0.326797 | 0.271605 | 0.102564  | 0.820513  | 0.254098  | 0.263158  | ...   | True      | False       | False      | True    | False  | False   | False      | False  | True    | True       |
|     621332 | 0.688312 | 0.50 | 1.000000 | 0.163399 | 0.111111 | 0.282051  | 0.256410  | 0.393443  | 0.500000  | ...   | True      | False       | False      | False   | False  | False   | False      | False  | True    | True       |
|     481293 | 0.701299 | 0.00 | 0.935622 | 0.169935 | 0.160494 | 0.487179  | 0.128205  | 0.188525  | 0.236842  | ...   | False     | True        | False      | True    | False  | False   | False      | False  | True    | True       |
|     297893 | 0.805195 | 0.50 | 0.373391 | 0.522876 | 0.456790 | 0.487179  | 1.000000  | 0.000000  | 0.000000  | ...   | False     | True        | True       | True    | True   | True    | False      | True   | False   | True       |
|     615697 | 0.000000 | 0.00 | 0.103004 | 0.372549 | 0.074074 | 0.615385  | 0.615385  | 0.352459  | 0.368421  | ...   | False     | False       | True       | False   | False  | False   | False      | False  | True    | True       |
|     337834 | 0.545455 | 0.25 | 0.201717 | 0.300654 | 0.419753 | 0.000000  | 0.000000  | 0.459016  | 0.421053  | ...   | False     | True        | False      | True    | False  | False   | False      | True   | True    | True       |
|     843362 | 0.844156 | 0.00 | 0.206009 | 0.751634 | 0.604938 | 0.641026  | 0.615385  | 0.475410  | 0.500000  | ...   | True      | False       | False      | True    | True   | False   | False      | True   | False   | True       |
|     995177 | 0.818182 | 0.25 | 0.618026 | 0.562092 | 0.728395 | 0.230769  | 0.358974  | 0.311475  | 0.315789  | ...   | False     | True        | True       | True    | True   | True    | True       | True   | True    | True       |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [368]:
# Code to rename all the columns in the dataset
combined2 = combined
column_renaming = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'class': 'class'
}


combined2.rename(columns=column_renaming, inplace=True)
print(combined2.head())


   unique_id       age  blood pressure  blood glucose random  blood urea  \
0     474407  0.194805            0.75              0.158798    0.196078   
1     137148  0.870130            1.00              0.965665    0.522876   
2     484175  0.610390            0.75              0.000000    0.633987   
3     343710  0.714286            0.50              0.442060    0.901961   
4     532520  0.701299            0.75              0.150215    0.281046   

   serum creatinine    sodium  potassium  hemoglobin  packed cell volume  ...  \
0          0.160494  0.358974   0.256410    0.221311            0.184211  ...   
1          0.641975  0.743590   0.102564    0.295082            0.368421  ...   
2          0.839506  0.076923   0.307692    0.319672            0.342105  ...   
3          0.432099  0.615385   0.692308    0.172131            0.210526  ...   
4          0.234568  0.641026   0.692308    0.450820            0.447368  ...   

   pc_normal  pcc_present  ba_present  htn_yes  dm_yes  

/var/folders/w0/0vsxrnfs6cg802b64pp6tf2h0000gn/T/ipykernel_64659/1610535754.py:32: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  combined2.rename(columns=column_renaming, inplace=True)


## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [369]:
# Converted SQL to Pandas code
result = combined2.groupby('Target_ckd').size().reset_index(name='count')
print(result)

   Target_ckd  count
0       False    111
1        True     26


*Describe what the above code does here*
The code above is grouping by Target_ckd which means all the rows in Target_ckd with the same value are grouped together. Then it counts the number of those values in each row. Then it converts the grouped result into a data frame and renames the count column to "count".
